solvers/lcs_factory.cc

```diff
@@ -281,62 +301,104 @@ LCS LCSFactory::LinearizePlantToLCS(
     const drake::systems::Context<double>& context,
     const std::vector<drake::SortedPair<drake::geometry::GeometryId>>&
         contact_geoms,
     int num_friction_directions, const std::vector<double>& mu,
-    dairlib::solvers::ContactModel contact_model) {
+    dairlib::solvers::ContactModel contact_model, bool with_z_lambda) {
   int n_contacts = contact_geoms.size();
 
   int n_v = plant.num_velocities();
 
+  vector<int> num_direction_contacts_vector;
+  int num_planar_contacts = 0;
+  for (auto pair : contact_geoms) {
+    if (IsObjToObjContact(plant,context,pair) && (!with_z_lambda)) {
+      num_planar_contacts++;
+      num_direction_contacts_vector.push_back(1);
+    }else {
+      num_direction_contacts_vector.push_back(num_friction_directions);
+    }
+  }
+
+  int num_direction_contacts = with_z_lambda? 2 * n_contacts * num_friction_directions : 2*num_friction_directions*(n_contacts - num_planar_contacts) + 2 * 1 * num_planar_contacts;
+
+
   VectorXd phi(n_contacts);
   MatrixXd J_n(n_contacts, n_v);
-  MatrixXd J_t(2 * n_contacts * num_friction_directions, n_v);
+  MatrixXd J_t(num_direction_contacts, n_v);
   std::vector<VectorXd> contact_points;
+
   for (int i = 0; i < n_contacts; i++) {
-    multibody::GeomGeomCollider collider(plant, contact_geoms[i]);
-    auto [phi_i, J_i] = collider.EvalPolytope(context, num_friction_directions);
-    auto [p_WCa, p_WCb] = collider.CalcWitnessPoints(context);
-    // TODO(yangwill): think about if we want to push back both witness points
-    contact_points.push_back(p_WCa);
-    phi(i) = phi_i;
-    J_n.row(i) = J_i.row(0);
-    J_t.block(2 * i * num_friction_directions, 0, 2 * num_friction_directions,
-              n_v) = J_i.block(1, 0, 2 * num_friction_directions, n_v);
+    multibody::GeomGeomCollider collider(
+        plant,
+        contact_geoms[i]);
+
+    if (num_direction_contacts_vector[i] == 1) {
+      Eigen::Vector3d planar_normal;
+      planar_normal << 0, 0, 1;
+      auto [phi_i, J_i] = collider.EvalPlanar(context, planar_normal);
+      auto [p_WCa, p_WCb] = collider.CalcWitnessPoints(context);
+      contact_points.push_back(p_WCa);
+      phi(i) = phi_i;
+      J_n.row(i) = J_i.row(0);
+      J_t.block(2 * std::accumulate(num_direction_contacts_vector.begin(),num_direction_contacts_vector.begin()+i,0), 0, 2 * num_direction_contacts_vector[i],
+                n_v) = J_i.block(1, 0, 2 * num_direction_contacts_vector[i], n_v);
+
+    } else {
+      auto [phi_i, J_i] =
+          collider.EvalPolytope(context, num_direction_contacts_vector[i]);
+      auto [p_WCa, p_WCb] = collider.CalcWitnessPoints(context);
+      contact_points.push_back(p_WCa);
+      phi(i) = phi_i;
+      J_n.row(i) = J_i.row(0);
+      J_t.block(2 * std::accumulate(num_direction_contacts_vector.begin(),num_direction_contacts_vector.begin()+i,0), 0, 2 * num_direction_contacts_vector[i],
+                n_v) = J_i.block(1, 0, 2 * num_direction_contacts_vector[i], n_v);
+    }
```

Comment:
The only difference between the `if` and `else` is which normal to use.  Suggest structure like:
```
if (num_direciton_contacts_vector[i] == 1) { // Do EvalPlanar }
else { // Do EvalPolytope }
// Then after the if/else do all the other commands, starting with collider.CalcWitnessPoints
```